### ISSUE TITLE
Why motion is represented by velocity, location, rotation to be training data?

### ISSUE THREAD

**Issue by hdjsjyl** [LinghaoChan/UniMoCap#3]
Hi Author,
    I have gone through the raw_pose_processing and understand all the steps it is doing.
    But I don't get point of motion_representation.
    Would you generally explain what motion_representation?
   And also, like the title, Why motion is represented by velocity, location, rotation to be training data?
   Thank you so much!

**Comment by LinghaoChan** [comment]
Hi @hdjsjyl 

We extend the motion representation from [HumanML3D](https://openaccess.thecvf.com/content/CVPR2022/html/Guo_Generating_Diverse_and_Natural_3D_Human_Motions_From_Text_CVPR_2022_paper.html), which is a redundant representation.

**Comment by hdjsjyl** [comment]
@LinghaoChan Thanks for your reply. I remember you asked the similar question in HumanML3D repo and you have a note for the motion representation code. Is it possible to share your notes with me about motion representation?

**Comment by LinghaoChan** [comment]
> @LinghaoChan Thanks for your reply. I remember you asked the similar question in HumanML3D repo and you have a note for the motion representation code. Is it possible to share your notes with me about motion representation?

What's the note you mentioned? I am not clear about it. And which issue in HumanML3D?

**Comment by hdjsjyl** [comment]
Hi @LinghaoChan ,
     You opened this question in the HumanML3D repo. https://github.com/EricGuo5513/HumanML3D/issues/74#issue-1802437427     In your questions, I saw you put some comments to the code. I think that is very good for understanding the code. 
![image](https://github.com/LinghaoChan/UniMoCap/assets/7590139/adbeaecf-fb1d-4700-be28-3bf7aded5a09)

Is it possible to share? Really appreciate your reply. Thanks

**Comment by LinghaoChan** [comment]
@hdjsjyl You can find it here. [link](https://github.com/LinghaoChan/UniMoCap/blob/b438feafadb9c4863058c78286e54a994f142181/motion_representation.py#L305)

**Comment by hdjsjyl** [comment]
Thank you.

**Comment by hdjsjyl** [comment]
@LinghaoChan hi linhao, Before, you said you have a plan to release a TMR pretrained model on MotionX dataset. I would kindly ask when you have the plan to release the model?

**Comment by LinghaoChan** [comment]
1-2 weeks perhaps. This is my work in recent weeks.

**Comment by hdjsjyl** [comment]
Hi @LinghaoChan , really appreciate your reply and help. Looking forward to the model. 
If you need help from my side about data or cope process. Pls let me know. 
Thank you.

**Comment by hdjsjyl** [comment]
Hi @LinghaoChan , another two quick questions: 1. Would you release any motion generation model trained on MotionX? 2. Would you release the training data of MotionX you used for training TMR? Thanks

**Comment by LinghaoChan** [comment]
> Hi @LinghaoChan , another two quick questions: 1. Would you release any motion generation model trained on MotionX? 2. Would you release the training data of MotionX you used for training TMR? Thanks

The first one might be a bit longer. The second one is yes. I am working on cleaning the codes. I just wake up. I will share you a  demo link of TMR when I'm at office.

**Comment by hdjsjyl** [comment]
Hi @LinghaoChan , thanks for your reply and demo link. If you need my help for any code cleaning or data cleaning, pls let me know. Thank you so much!

**Comment by LinghaoChan** [comment]
@hdjsjyl Plz check this demo link. https://swanhub.co/demo/Evan/OpenTMR

Any email or other ways I can reach out u?

**Comment by hdjsjyl** [comment]
Hi @LinghaoChan , this demo is very cool. My email is hdjsjyl@gmail.com, my wechat is hourlei. Welcome to connect with you, thank you.

**Comment by 2019211753** [comment]
@LinghaoChan I wonder if the root_y_position is a relative position with respect to the pelvis of the preceding frame, or if it is a global absolute position? Thank you in advance for your reply!

**Comment by LinghaoChan** [comment]
> @LinghaoChan I wonder if the root_y_position is a relative position with respect to the pelvis of the preceding frame, or if it is a global absolute position? Thank you in advance for your reply!

yes, the previous one.

**Comment by 2019211753** [comment]
> > @LinghaoChan I wonder if the root_y_position is a relative position with respect to the pelvis of the preceding frame, or if it is a global absolute position? Thank you in advance for your reply!
> 
> yes, the previous one.

Thanks for your rapid reply! Could I ask another question? Since it's a totally relative representation, what does the 'root_y_position' of the first frame refer to? I debug the code and find that it's not zero.

**Comment by LinghaoChan** [comment]
> > > @LinghaoChan I wonder if the root_y_position is a relative position with respect to the pelvis of the preceding frame, or if it is a global absolute position? Thank you in advance for your reply!
> > 
> > 
> > yes, the previous one.
> 
> Thanks for your rapid reply! Could I ask another question? Since it's a totally relative representation, what does the 'root_y_position' of the first frame refer to? I debug the code and find that it's not zero.

no explicit meaning. I do not remember which of the first or the final frame value is not used. But only T-1 values are used.

**Comment by 2019211753** [comment]
> > > > @LinghaoChan I wonder if the root_y_position is a relative position with respect to the pelvis of the preceding frame, or if it is a global absolute position? Thank you in advance for your reply!
> > > 
> > > 
> > > yes, the previous one.
> > 
> > 
> > Thanks for your rapid reply! Could I ask another question? Since it's a totally relative representation, what does the 'root_y_position' of the first frame refer to? I debug the code and find that it's not zero.
> 
> no explicit meaning. I do not remember which of the first or the final frame value is not used. But only T-1 values are used.

I really appreciate your informative explanation!🥰

**Comment by 2019211753** [comment]
> > @LinghaoChan I wonder if the root_y_position is a relative position with respect to the pelvis of the preceding frame, or if it is a global absolute position? Thank you in advance for your reply!
> 
> yes, the previous one.



> > @LinghaoChan I wonder if the root_y_position is a relative position with respect to the pelvis of the preceding frame, or if it is a global absolute position? Thank you in advance for your reply!
> 
> yes, the previous one.



> > @LinghaoChan I wonder if the root_y_position is a relative position with respect to the pelvis of the preceding frame, or if it is a global absolute position? Thank you in advance for your reply!
> 
> yes, the previous one.

I'm sorry for bother you again. I checked the recover_root_rot_pos function:
```python
def recover_root_rot_pos(data):
    rot_vel = data[..., 0]
    r_rot_ang = torch.zeros_like(rot_vel).to(data.device)
    '''Get Y-axis rotation from rotation velocity'''
    r_rot_ang[..., 1:] = rot_vel[..., :-1]
    r_rot_ang = torch.cumsum(r_rot_ang, dim=-1)

    r_rot_quat = torch.zeros(data.shape[:-1] + (4,)).to(data.device)
    r_rot_quat[..., 0] = torch.cos(r_rot_ang)
    r_rot_quat[..., 2] = torch.sin(r_rot_ang)

    r_pos = torch.zeros(data.shape[:-1] + (3,)).to(data.device)
    r_pos[..., 1:, [0, 2]] = data[..., :-1, 1:3]
    '''Add Y-axis rotation to root position'''
    r_pos = qrot(qinv(r_rot_quat), r_pos)

    r_pos = torch.cumsum(r_pos, dim=-2)

    r_pos[..., 1] = data[..., 3]
    return r_rot_quat, r_pos
```
I find that on the line: 
r_pos[..., 1] = data[..., 3]
It seems that it assign the root_y_position to r_pos directly without the cumsum operation. It is right?

**Comment by LinghaoChan** [comment]
> > > @LinghaoChan I wonder if the root_y_position is a relative position with respect to the pelvis of the preceding frame, or if it is a global absolute position? Thank you in advance for your reply!
> > 
> > 
> > yes, the previous one.
> 
> > > @LinghaoChan I wonder if the root_y_position is a relative position with respect to the pelvis of the preceding frame, or if it is a global absolute position? Thank you in advance for your reply!
> > 
> > 
> > yes, the previous one.
> 
> > > @LinghaoChan I wonder if the root_y_position is a relative position with respect to the pelvis of the preceding frame, or if it is a global absolute position? Thank you in advance for your reply!
> > 
> > 
> > yes, the previous one.
> 
> I'm sorry for bother you again. I checked the recover_root_rot_pos function:
> 
> ```python
> def recover_root_rot_pos(data):
>     rot_vel = data[..., 0]
>     r_rot_ang = torch.zeros_like(rot_vel).to(data.device)
>     '''Get Y-axis rotation from rotation velocity'''
>     r_rot_ang[..., 1:] = rot_vel[..., :-1]
>     r_rot_ang = torch.cumsum(r_rot_ang, dim=-1)
> 
>     r_rot_quat = torch.zeros(data.shape[:-1] + (4,)).to(data.device)
>     r_rot_quat[..., 0] = torch.cos(r_rot_ang)
>     r_rot_quat[..., 2] = torch.sin(r_rot_ang)
> 
>     r_pos = torch.zeros(data.shape[:-1] + (3,)).to(data.device)
>     r_pos[..., 1:, [0, 2]] = data[..., :-1, 1:3]
>     '''Add Y-axis rotation to root position'''
>     r_pos = qrot(qinv(r_rot_quat), r_pos)
> 
>     r_pos = torch.cumsum(r_pos, dim=-2)
> 
>     r_pos[..., 1] = data[..., 3]
>     return r_rot_quat, r_pos
> ```
> 
> I find that on the line: r_pos[..., 1] = data[..., 3] It seems that it assign the root_y_position to r_pos directly without the cumsum operation. It is right?

@2019211753 Sorry for my confusing statement before. I'm such a goldfish brain. I only remember the y axis is different from others and forget the details due to the too-long history. The root representation is the root rotation velocity, two linear velocities, and the y position. The overall root representation is 4-dim. 

Please visit the representation annotation [here](https://github.com/LinghaoChan/UniMoCap/blob/25cb2fc65749cb7a03f9db79eb1bfcd4e0642409/motion_representation.py#L308).

**Comment by 2019211753** [comment]
> > > > @LinghaoChan I wonder if the root_y_position is a relative position with respect to the pelvis of the preceding frame, or if it is a global absolute position? Thank you in advance for your reply!
> > > 
> > > 
> > > yes, the previous one.
> > 
> > 
> > > > @LinghaoChan I wonder if the root_y_position is a relative position with respect to the pelvis of the preceding frame, or if it is a global absolute position? Thank you in advance for your reply!
> > > 
> > > 
> > > yes, the previous one.
> > 
> > 
> > > > @LinghaoChan I wonder if the root_y_position is a relative position with respect to the pelvis of the preceding frame, or if it is a global absolute position? Thank you in advance for your reply!
> > > 
> > > 
> > > yes, the previous one.
> > 
> > 
> > I'm sorry for bother you again. I checked the recover_root_rot_pos function:
> > ```python
> > def recover_root_rot_pos(data):
> >     rot_vel = data[..., 0]
> >     r_rot_ang = torch.zeros_like(rot_vel).to(data.device)
> >     '''Get Y-axis rotation from rotation velocity'''
> >     r_rot_ang[..., 1:] = rot_vel[..., :-1]
> >     r_rot_ang = torch.cumsum(r_rot_ang, dim=-1)
> > 
> >     r_rot_quat = torch.zeros(data.shape[:-1] + (4,)).to(data.device)
> >     r_rot_quat[..., 0] = torch.cos(r_rot_ang)
> >     r_rot_quat[..., 2] = torch.sin(r_rot_ang)
> > 
> >     r_pos = torch.zeros(data.shape[:-1] + (3,)).to(data.device)
> >     r_pos[..., 1:, [0, 2]] = data[..., :-1, 1:3]
> >     '''Add Y-axis rotation to root position'''
> >     r_pos = qrot(qinv(r_rot_quat), r_pos)
> > 
> >     r_pos = torch.cumsum(r_pos, dim=-2)
> > 
> >     r_pos[..., 1] = data[..., 3]
> >     return r_rot_quat, r_pos
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I find that on the line: r_pos[..., 1] = data[..., 3] It seems that it assign the root_y_position to r_pos directly without the cumsum operation. It is right?
> 
> @2019211753 Sorry for my confusing statement before. I'm such a goldfish brain. I only remember the y axis is different from others and forget the details due to the too-long history. The root representation is the root rotation velocity, two linear velocities, and the y position. The overall root representation is 4-dim.
> 
> Please visit the representation annotation [here](https://github.com/LinghaoChan/UniMoCap/blob/25cb2fc65749cb7a03f9db79eb1bfcd4e0642409/motion_representation.py#L308).

Thanks again for your reply. Then,  could I say the y-axis represents the global position?

**Comment by LinghaoChan** [comment]
> > > > > @LinghaoChan I wonder if the root_y_position is a relative position with respect to the pelvis of the preceding frame, or if it is a global absolute position? Thank you in advance for your reply!
> > > > 
> > > > 
> > > > yes, the previous one.
> > > 
> > > 
> > > > > @LinghaoChan I wonder if the root_y_position is a relative position with respect to the pelvis of the preceding frame, or if it is a global absolute position? Thank you in advance for your reply!
> > > > 
> > > > 
> > > > yes, the previous one.
> > > 
> > > 
> > > > > @LinghaoChan I wonder if the root_y_position is a relative position with respect to the pelvis of the preceding frame, or if it is a global absolute position? Thank you in advance for your reply!
> > > > 
> > > > 
> > > > yes, the previous one.
> > > 
> > > 
> > > I'm sorry for bother you again. I checked the recover_root_rot_pos function:
> > > ```python
> > > def recover_root_rot_pos(data):
> > >     rot_vel = data[..., 0]
> > >     r_rot_ang = torch.zeros_like(rot_vel).to(data.device)
> > >     '''Get Y-axis rotation from rotation velocity'''
> > >     r_rot_ang[..., 1:] = rot_vel[..., :-1]
> > >     r_rot_ang = torch.cumsum(r_rot_ang, dim=-1)
> > > 
> > >     r_rot_quat = torch.zeros(data.shape[:-1] + (4,)).to(data.device)
> > >     r_rot_quat[..., 0] = torch.cos(r_rot_ang)
> > >     r_rot_quat[..., 2] = torch.sin(r_rot_ang)
> > > 
> > >     r_pos = torch.zeros(data.shape[:-1] + (3,)).to(data.device)
> > >     r_pos[..., 1:, [0, 2]] = data[..., :-1, 1:3]
> > >     '''Add Y-axis rotation to root position'''
> > >     r_pos = qrot(qinv(r_rot_quat), r_pos)
> > > 
> > >     r_pos = torch.cumsum(r_pos, dim=-2)
> > > 
> > >     r_pos[..., 1] = data[..., 3]
> > >     return r_rot_quat, r_pos
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I find that on the line: r_pos[..., 1] = data[..., 3] It seems that it assign the root_y_position to r_pos directly without the cumsum operation. It is right?
> > 
> > 
> > @2019211753 Sorry for my confusing statement before. I'm such a goldfish brain. I only remember the y axis is different from others and forget the details due to the too-long history. The root representation is the root rotation velocity, two linear velocities, and the y position. The overall root representation is 4-dim.
> > Please visit the representation annotation [here](https://github.com/LinghaoChan/UniMoCap/blob/25cb2fc65749cb7a03f9db79eb1bfcd4e0642409/motion_representation.py#L308).
> 
> Thanks again for your reply. Then, could I say the y-axis represents the global position?

Yep. You can visualize it for double checking.

**Comment by 2019211753** [comment]
> > > > > > @LinghaoChan I wonder if the root_y_position is a relative position with respect to the pelvis of the preceding frame, or if it is a global absolute position? Thank you in advance for your reply!
> > > > > 
> > > > > 
> > > > > yes, the previous one.
> > > > 
> > > > 
> > > > > > @LinghaoChan I wonder if the root_y_position is a relative position with respect to the pelvis of the preceding frame, or if it is a global absolute position? Thank you in advance for your reply!
> > > > > 
> > > > > 
> > > > > yes, the previous one.
> > > > 
> > > > 
> > > > > > @LinghaoChan I wonder if the root_y_position is a relative position with respect to the pelvis of the preceding frame, or if it is a global absolute position? Thank you in advance for your reply!
> > > > > 
> > > > > 
> > > > > yes, the previous one.
> > > > 
> > > > 
> > > > I'm sorry for bother you again. I checked the recover_root_rot_pos function:
> > > > ```python
> > > > def recover_root_rot_pos(data):
> > > >     rot_vel = data[..., 0]
> > > >     r_rot_ang = torch.zeros_like(rot_vel).to(data.device)
> > > >     '''Get Y-axis rotation from rotation velocity'''
> > > >     r_rot_ang[..., 1:] = rot_vel[..., :-1]
> > > >     r_rot_ang = torch.cumsum(r_rot_ang, dim=-1)
> > > > 
> > > >     r_rot_quat = torch.zeros(data.shape[:-1] + (4,)).to(data.device)
> > > >     r_rot_quat[..., 0] = torch.cos(r_rot_ang)
> > > >     r_rot_quat[..., 2] = torch.sin(r_rot_ang)
> > > > 
> > > >     r_pos = torch.zeros(data.shape[:-1] + (3,)).to(data.device)
> > > >     r_pos[..., 1:, [0, 2]] = data[..., :-1, 1:3]
> > > >     '''Add Y-axis rotation to root position'''
> > > >     r_pos = qrot(qinv(r_rot_quat), r_pos)
> > > > 
> > > >     r_pos = torch.cumsum(r_pos, dim=-2)
> > > > 
> > > >     r_pos[..., 1] = data[..., 3]
> > > >     return r_rot_quat, r_pos
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > I find that on the line: r_pos[..., 1] = data[..., 3] It seems that it assign the root_y_position to r_pos directly without the cumsum operation. It is right?
> > > 
> > > 
> > > @2019211753 Sorry for my confusing statement before. I'm such a goldfish brain. I only remember the y axis is different from others and forget the details due to the too-long history. The root representation is the root rotation velocity, two linear velocities, and the y position. The overall root representation is 4-dim.
> > > Please visit the representation annotation [here](https://github.com/LinghaoChan/UniMoCap/blob/25cb2fc65749cb7a03f9db79eb1bfcd4e0642409/motion_representation.py#L308).
> > 
> > 
> > Thanks again for your reply. Then, could I say the y-axis represents the global position?
> 
> Yep. You can visualize it for double checking.

Thanks😊
